pep-0xxx.rst

```diff
@@ -0,0 +1,1317 @@
+two-octet cipher identifier as used in the TLS handshake, stored as a tuple of
+integers. The names of the enum members will be their IANA-registered cipher
+suite names.
+
+Rather than populate this enum by hand, it is likely that we'll define a
+script that can build it from Christian Heimes' `tlsdb JSON file`_ (warning:
+large file). This also opens up the possibility of extending the API with
+additional querying function, such as determining which TLS versions support
+which ciphers, if that functionality is found to be useful or necessary.
+
+If users find this approach to be onerous, a future extension to this API can
+provide helpers that can reintroduce OpenSSL's aggregation functionality.
+
+Because this enum would be enormous, the entire enum is not provided here.
+Instead, a small sample of entries is provided to give a flavor of how it will
+appear.
+
+::
+
+    class CipherSuite(Enum):
+        ...
+        TLS_ECDHE_RSA_WITH_3DES_EDE_CBC_SHA = (0xC0, 0x12)
+        ...
+        TLS_ECDHE_ECDSA_WITH_AES_128_CCM = (0xC0, 0xAC)
+        ...
+        TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256 = (0xC0, 0x2B)
+        ...
+
+
+Protocol Negotiation
+~~~~~~~~~~~~~~~~~~~~
+
+Both NPN and ALPN allow for protocol negotiation as part of the HTTP/2
+handshake. While NPN and ALPN are, at their fundamental level, built on top of
+bytestrings, string-based APIs are frequently problematic as they allow for
+errors in typing that can be hard to detect.
+
+For this reason, this module would define a type that protocol negotiation
+implementations can pass and be passed. This type would wrap a bytestring to
+allow for aliases for well-known protocols. This allows us to avoid the
+problems inherent in typos for well-known protocols, while allowing the full
+extensibility of the protocol negotiation layer if needed by letting users pass
+byte strings directly.
+
+::
+
+    class NextProtocol(Enum):
+        H2 = b'h2'
+        H2C = b'h2c'
+        HTTP1 = b'http/1.1'
+        WEBRTC = b'webrtc'
+        C_WEBRTC = b'c-webrtc'
+        FTP = b'ftp'
+        STUN = b'stun.nat-discovery'
+        TURN = b'stun.turn'
+
+TLS Versions
+~~~~~~~~~~~~
+
+It is often useful to be able to restrict the versions of TLS you're willing to
+support. There are many security advantages in refusing to use old versions of
+TLS, and some misbehaving servers will mishandle TLS clients advertising
+support for newer versions.
+
+The following enumerated type can be used to gate TLS versions. Forward-looking
+applications should almost never set a maximum TLS version unless they
+absolutely must, as a TLS backend that is newer than the Python that uses it
+may support TLS versions that are not in this enumerated type.
+
+Additionally, this enumerated type defines two additional flags that can always
+be used to request either the lowest or highest TLS version supported by an
+implementation.
+
+::
+
+    class TLSVersion(Enum):
```

Comment:
I have two comments about TLSVersion. First, I would rather see ```TLSv1_0``` instead of ```TLSv1``` to make the version more explicit. Second, can we use the on-wire constants (e.g. ```0x0304``` for TLS 1.3) instead of auto-generated enum values? The PEP already uses constants for ciphers so it may make sense to use known constants for TLS versions, too.